### ISSUE TITLE
.values on ExtensionArray-backed containers

### ISSUE THREAD

**Issue by TomAugspurger** [pandas-dev/pandas#19954]
Discussed briefly on the call today, but we should go through things formally.

What should the return type of `Series[extension_array].values` and `Index[extension_array].values` be? I believe the two options are

1. Return the ExtensionArray backing it (e.g. like what Categorical does)
2. Return an ndarray with some information loss / performance cost
   - e.g. like Series[datetimeTZ].values -> datetime64ns at UTC
   - e.g. Series[period].values -> ndarray[Period objects]

## Current State

Not sure how much weight we should put on the current behavior, but for reference:

type        | Series.values           | Index.values
----------- | ----------------------- | ------------
datetime    | datetime64ns            | datetime64ns
datetime-tz | datetine64ns(UTC&naive) | datetime64ns(UTC&naive)
categorical | Categorical             | Categorical
period      | NA                      | ndarray[Period objects]
interval    | NA                      | ndarray[Interval objects]

<details>

```python
In [5]: pd.Series(pd.date_range('2017', periods=1)).values
Out[5]: array(['2017-01-01T00:00:00.000000000'], dtype='datetime64[ns]')

In [6]: pd.Series(pd.date_range('2017', periods=1, tz='US/Eastern')).values
Out[6]: array(['2017-01-01T05:00:00.000000000'], dtype='datetime64[ns]')

In [7]: pd.Series(pd.Categorical([1])).values
Out[7]:
[1]
Categories (1, int64): [1]

In [8]: pd.Series(pd.SparseArray([1])).values
Out[8]:
[1]
Fill: 0
IntIndex
Indices: array([0], dtype=int32)

In [9]: pd.date_range('2017', periods=1).values
Out[9]: array(['2017-01-01T00:00:00.000000000'], dtype='datetime64[ns]')

In [10]: pd.date_range('2017', periods=1, tz='US/Central').values
Out[10]: array(['2017-01-01T06:00:00.000000000'], dtype='datetime64[ns]')

In [11]: pd.period_range('2017', periods=1, freq='D').values
Out[11]: array([Period('2017-01-01', 'D')], dtype=object)

In [12]: pd.interval_range(start=0, periods=1).values
Out[12]: array([Interval(0, 1, closed='right')], dtype=object)

In [13]: pd.CategoricalIndex([1]).values
Out[13]:
[1]
Categories (1, int64): [1]
```

</details>

If we decide to have the return values be ExtensionArrays, we'll need to discuss
to what extent they're part of the public API.

Regardless of the choice for `.values`, we'll probably want to support the other
use case (maybe just by documenting "call `np.asarray` on it). Internally, we
have `._values` ("best" array, ndarray or EA) and `._ndarray_values` (always an
ndarray).

cc @jreback @jorisvandenbossche @jschendel @jbrockmendel @shoyer @chris-b1

**Comment by TomAugspurger** [comment]
Not to derail the discussion from the get-go, but there's a closely related topic of how many "nuisance" ndarray methods we throw on the ExtensionArray base class to make writing code that's ndarray vs. extensionarray agnostic easier. For example `tolist`, which just calls `list(self)`. `.reshape`, which just returns `self`. The list goes on (and on).

**Comment by shoyer** [comment]
I was pretty happy with always returning an `ndarray` from `.values`, but we broke that a long time ago with `Categorical`. For consistency we should probably clean this up (even for existing extension arrays), but unfortunately this will be a painful compatibility break.

I think my preferred option would be to make `obj.values` always return the underlying `ExtensionArray`, guaranteed to be the same data without any copies (so it can be modified in place). For converting to NumPy, we should probably add a `obj.to_numpy()` method on Series/DataFrame that is basically sugar for `np.asarray(obj)`

> datetine64ns(UTC&naive)

NumPy's datetime64 is now always timezone naive, not UTC. I guess you meant the times are converted into UTC before being passed into NumPy?

> there's a closely related topic of how many "nuisance" ndarray methods we throw on the ExtensionArray base class to make writing code that's ndarray vs. extensionarray agnostic easier

My preference would be to make the base class as abstract as possible, only implementing methods that are actually used by pandas. This will minimize the effort required to author new ExtensionArray classes, rather than requiring that authors override pre-existing methods.  Extension arrays should be a low-level interface -- we already have the high-level interface for extension arrays in Series/DataFrame.

**Comment by chris-b1** [comment]
> I think my preferred option would be to make obj.values always return the underlying ExtensionArray, guaranteed to be the same data without any copies (so it can be modified in place). For converting to NumPy, we should probably add a obj.to_numpy() method on Series/DataFrame that is basically sugar for np.asarray(obj)

+1 on this.  I think the break in compatibility is less painful than the inconsistency and current situation of, e.g. `PeriodIndex.values` having very surprising performance costs.

xref to #15750 - was about changing the return type of tz-aware datetimes to boxed scalars.

**Comment by TomAugspurger** [comment]
>  I guess you meant the times are converted into UTC before being passed into NumPy?

Yep

I think I share the preference for `.values` being the actual array backing the data (even though that's not always possible with `DataFrame.values`).

Also +1 for a `to_numpy` method on Series / Index / Frame, with the same copy vs. view behavior as `asarray`.

**Comment by TomAugspurger** [comment]
Are people OK with `Series[datetime64ns].values` being an ndarray, but `Series[datetime64ns, tz].values` an ExtensionArray? That seems potentially confusing.

**Comment by jreback** [comment]
it has. even this way for a long time
we should change the former to be honest - but last time there were some objections

**Comment by TomAugspurger** [comment]
> it has been this way for a long time

What do you mean by that?

>  we should change the former to be honest - but last time there were some objections

Why would we change tz-naive to return anything other than a datetime64[ns] ndarray?

**Comment by chris-b1** [comment]
As stated above I'm OK with the inconsistency, but see where it could be surprising. 

Another option would be to deprecate `.values` in favor of an explicit `to_numpy()` and another accessor that always returns the backing array, numpy or not, something like `Series.array`?

**Comment by TomAugspurger** [comment]
> As stated above I'm OK with the inconsistency, but see where it could be surprising.

👍 . I wanted to get some explicit thoughts on that one, since I didn't initially comprehend that two seemingly similar types could have different outcomes.

> another accessor that always returns the backing array

Yes. We'll want something like that. Internally, we've used `._values` and `._ndarray_values`.

**Comment by TomAugspurger** [comment]
Let's start with goals and work backwards to an API

1. I want the actual array that's backing this Series / Index.
   I cannot afford a copy.
2. I want an ndarray holds the same values as the backing array.
   This ndarray may not be the actual one backing the Series / Index,
   and getting it may involve some data copying / coercion. I'm willing
   to lose some information (categorical types, timezone, etc.), but I
   need to be able to unambiguously identify a value (i.e. I want the
   category `'a'` not the code `0`. Not sure about tzinfo here...).
3. I want some kind of array that represents the data backing the array.
   I'd really like to avoid data copying / coercion (think Categorical.codes,
   PeriodArray.asi8, etc). I'm willing to lose information to avoid copying
   data unnecessarily.

I think getting 1 and 2 right are necessary. 3 would be a nice thing to have, but I wouldn't
consider it a blocker for 0.24.0.

I also think that "fixing" `.values` isn't something we can do at this point. The
current behavior is inconsistent between EAs, and adding new EAs would force us to
break API or break consistency with Categorical.

So, I propose avoiding that fight. Let's use new names to achieve the behavior we want.
This seems important enough to warrant the API bloat, and I think we can remove mentions
of `.values` in the docs, so we're net +2 properties here.

---

Here's a concrete proposal. I still need to go through all our EAs to figure out an
exact set of behavior (`datetime64[ns, tz]` may be particularly troublesome).

```python
@property
def array(self) -> Union[ndarray, ExtensionArray]:
    """Return the array backing this Series or Index

    Examples
    --------
    >>> arr = pd.core.arrays.period_array(['2000', '2001'], freq='A')
    >>> ser = pd.Series(arr)
    >>> ser.array
    <PeriodArray>
    ['2000', '2001']
    Length: 2, dtype: period[A-DEC]
    """
    return self._values


@property
def numpy_array(self) -> ndarray:
    """Return a NumPy array of this object's values.

    This may or may not require a copy or coercion of values.
    For dtypes that can be represented by NumPy, this will be a view on
    the actual values. For ExtensionArrays, this will likely be an object-dtype
    ndarray that losslessly represents the values.

    Examples
    --------
    >>> arr = pd.core.arrays.period_array(['2000', '2001'], freq='A')
    >>> ser = pd.Series(arr)
    >>> ser.numpy_array
    array([Period('2000', 'A-DEC'), Period('2001', 'A-DEC')], dtype=object)
    """
    return np.asarray(self._values)


@property
def ndarray_values(self) -> ndarray:
    """Return a NumPy array representing the values in this object.

    This should be faster to compute than ``self.numpy_array``, but will
    require additional context to interpret.
    Examples
    --------
    >>> arr = pd.core.arrays.period_array(['2000', '2001'], freq='A')
    >>> ser = pd.Series(arr)
    >>> ser.ndarray_values
    array([30, 31])
    """
    return self._ndarray_values
```

**Comment by shoyer** [comment]
@TomAugspurger I like this general idea, but let me suggest two refinements:
- `numpy_array` should be a `to_numpy()` method (because the NumPy array is sometimes created on the fly)
- Instead of `ndarray_values`, call it something slightly more descriptive -- maybe `raw_array` or `base_array`? This property also might not be defined on all extension arrays, e.g., if they are written on top of arrow.

**Comment by jreback** [comment]
why would this not be on top of an arrow backed array? the point of EA is to completely hide this detail

**Comment by TomAugspurger** [comment]
> numpy_array should be a to_numpy() method (because the NumPy array is sometimes created on the fly)

I was viewing this as a `.values` replacement, which is why I went with a property initially. But if I were designing this from scratch I would certainly make it a method... I'm not sure how to balance these two. I think right now a method would be preferable here, but I could go either way.

> Instead of ndarray_values, call it something slightly more descriptive -- maybe raw_array or base_array?

Mmm, yes that's a good point... I would need to look at this. In several places, like in our indexing engines, I think we really need an ndarray right now.

> why would this not be on top of an arrow backed array? the point of EA is to completely hide this detail

I'm not sure what you mean. Were you responding to my proposal or Stehpan's comments?

**Comment by jreback** [comment]
I was responding to stephan

an arrow backed extension arrow is nothing special as far as pandas is concerned

**Comment by shoyer** [comment]
> an arrow backed extension arrow is nothing special as far as pandas is concerned

You can turn an arrow backed extension arrow into a NumPy array (possibly with a copy) or you can access it as an extension array, but there isn't necessarily any equivalent of a "raw numpy array" underlying the values that you can directly mutate to change the values in a `pandas.Series`.

**Comment by jreback** [comment]
@shoyer of course but how does this actually matter - again the actual implementation of the EA is irrelevant, except when we ask it give us a raw set of values and even that is transparent

**Comment by TomAugspurger** [comment]
I think whether or not this has to be an ndarray depends on how it's (intended) to be used. Is the primary motivation

1. a fast (but maybe not zero copy) representation of the data as an ndarray, so that I can do ndarray-like things to it? Or,
2. A fast array-like representation of the data, no copies allowed.

I was proposing the first, hence the name `ndarray_values`. But perhaps the second is what people want? This one is kind of ambiguous.

Right now, we seem to use `IndexOpsMixin._ndarray_values` a lot in the indexing engines. I suspect that needs to be an ndarray, and not an Arrow array (or buffer), but I haven't looked.

**Comment by shoyer** [comment]
> 2. A fast array-like representation of the data, no copies allowed.

This is definitely the use-case I had in mind. This can be convenient, e.g., for cases where you want to modify the NumPy array in place.

Note that this is definitely possible for some but not all arrow based arrays. For example, if you have integers or floats without any missing values.

> 1. a fast (but maybe not zero copy) representation of the data as an ndarray, so that I can do ndarray-like things to it?

I think this is covered by `to_numpy()`?

**Comment by TomAugspurger** [comment]
> > 1. a fast (but maybe not zero copy) representation of the data as an ndarray, so that I can do ndarray-like things to it?

> I think this is covered by to_numpy()?

Not the "fast" requirement though. For e.g. PeriodIndex, `PeriodIndex.to_numpy()` would be an ndarray of objects, which is expensive to create.

My last argument in favor of `ndarray_values -> ndarray` over a `raw_array -> ArrayLike` is that the `.raw_array` use-case isn't clear to me. The return type of `.raw_array` would be deliberately broad (since ExtensionArray makes no claims about the actual physical storage). So use it, you would have to know something about what the actual type is in your case. And if you're doing something specific to your array type, why not do `series.array.<type-specific-thing>`?

The usage for `ndarray_values` is a bit clearer to me. If you have some algorithm that has to operate on a ndararys, you *may* be able to use the algorithm on `ndarray_values` and then reconstruct the real result using additional information that you've held onto separately (think factorize, sorting).

As a final note, I think this `ndarray_values` / `raw_array` is the fuzziest concept out of the 3, and should not necessarily block progression on `.to_numpy()` (method or property) and `.array`.

**Comment by shoyer** [comment]
> My last argument in favor of ndarray_values -> ndarray over a raw_array -> ArrayLike is that the .raw_array use-case isn't clear to me. The return type of .raw_array would be deliberately broad (since ExtensionArray makes no claims about the actual physical storage). So use it, you would have to know something about what the actual type is in your case. And if you're doing something specific to your array type, why not do series.array.<type-specific-thing>?

I totally agree here, I was just proposing a different name for the saw use-case:
```
@property
def raw_numpy_array_values(self) -> np.ndarray:
    # raises an error if the extension array doesn't support it
```

**Comment by jorisvandenbossche** [comment]
Personally I would leave the `ndarray_values` or `raw_numpy_array_values` idea out of the public API on the *containers*. This concept will inherently be inconsistent between EAs, as it will only be possible for some of the EAs. As @shoyer noted this concept might not be possible for certain arrow-backed EAs, but another typical example is our own IntervalArray which is backed by two ndarrays (left/right). 

Of course, as @shoyer mentions, we can raise an error for this attribute if there is no clear no-copy option available. But then a user would need to be aware of the specific data type/EA that is being used, so then why not use EA-specific API to do this? 
For example for this mentioned use case:

> If you have some algorithm that has to operate on a ndararys, you may be able to use the algorithm on ndarray_values and then reconstruct the real result using additional information that you've held onto separately (think factorize, sorting).

and if you have a PeriodArray. Then you know how a PeriodArray works: it consistent of integers and a freq, so you can use the public PeriodArray.asi8 to access those integers, do something with it, and then reconstruct a PeriodArray from the integers and the freq. 

I would say that each ExtensionArray that wants to enable such manipulation, can provide it's own access to the underlying data, because this will differ for each array anyhow? (eg for PeriodArray, you already need to know you need to keep track of freq to reconstruct it, so then you can also use a specific attribute to access the data?)  
(and of course, having a somewhat consistent API eg across all datetime-like arrays is certainly nice, but that is something we can ensure for our own EAs)

**Comment by jorisvandenbossche** [comment]
Regarding `to_numpy()`: I agree with the reasons mentioned above to have it as a method and not an attribute, but another reason for having it as a method is that we then can provide options. 
Eg, for datetime tz, the user could choose to get it as object array with tz retained, or as UTC-converted naive M8[ns].

**Comment by TomAugspurger** [comment]
Alrighty thanks Joris.

> we then can provide options.

What would the function signature be for your `Series.to_numpy`? `**kwargs`? Or would pandas dictate what's allowed there?

I'll get started on a PR implementing this.

```python
@property
def array(self) -> Union[ndarray, ExtensionArray]:
    """Return the array backing this Series or Index

    Examples
    --------
    >>> arr = pd.core.arrays.period_array(['2000', '2001'], freq='A')
    >>> ser = pd.Series(arr)
    >>> ser.array
    <PeriodArray>
    ['2000', '2001']
    Length: 2, dtype: period[A-DEC]
    """
    return self._values


@property
def to_numpy(self) -> ndarray:
    """Return a NumPy array of this object's values.

    This may or may not require a copy or coercion of values.
    For dtypes that can be represented by NumPy, this will be a view on
    the actual values. For ExtensionArrays, this will likely be an object-dtype
    ndarray. Equality will be preserved (values in the array that were equal
    before will be equal in the ndarray; likewise for values that were unequal.
    However, the dtype information may be lost (e.g. categories for a Categorical).

    Examples
    --------
    >>> arr = pd.core.arrays.period_array(['2000', '2001'], freq='A')
    >>> ser = pd.Series(arr)
    >>> ser.numpy_array
    array([Period('2000', 'A-DEC'), Period('2001', 'A-DEC')], dtype=object)
    """
    return np.asarray(self._values)
```

**Comment by TomAugspurger** [comment]
What to do about datetime and datetimetz (cc @jbrockmendel)? Here's a proposal.


dtype              | array                    | to_numpy
------------------ | -------------------------| --------
datetime64[ns]     | DatetimeArray            | ndarray[datetime64[ns]] (view)
datetime64[ns, tz] | DatetimeArray            | ndarray[object] (Timestamp)
timedelta64[ns]    | ndarray[timedelta64[ns]] | ndarray[timedelta64[ns]] (view)


Alternatives:

- `Series[datetime64[ns]].array` could return an `ndarray[datetime64[ns]]`. But that's inconsistent with `.array` for `datetime64[ns, tz]`, and you have `to_numpy()` if you want that. I don't know which is better here.
- `Series[datetime64[ns, tz]].to_numpy()` can avoid conversion to object by converting to UTC and stripping timezone info. That would satisfy this "equality preservation" requirement I have in mind for `to_numpy` (things that were equal before will continue to be equal after `to_numpy()`. same for not equal).

**Comment by shoyer** [comment]
> Personally I would leave the ndarray_values or raw_numpy_array_values idea out of the public API on the containers. This concept will inherently be inconsistent between EAs, as it will only be possible for some of the EAs.

Sounds good to me. `.array` and `.to_numpy()` seems like enough.

**Comment by TomAugspurger** [comment]
Uses of ndarray_values

- Index.itemsize (deprecated)
- Index.strides (deprecated)
- Index._engine
- Index set ops
- Index.insert
- DatetimeIndex.unique
- MultiIndex.equals
- pytables._convert_index (shared across integer and period)

There are a few other uses (mostly datetime / timedelta / period) that could maybe uses asi8 instead. I'm not familiar enough with indexing to know whether that can operate on something other than ndarrays. In theory, EAs can implement the buffer protocol, which would get the data to cython. But I don't know what ops would be required when we're down there.

> Is anyone using PeriodArray._ndarray_values or DatetimeTZ._ndarray_values? (i.e. the lossy ones)

Yes, when writing to HDF5.

> Is the proposal to implement to_numpy in addition to _ndarray_values or to replace it? (with deprecation cycle where relevant)

`to_numpy` and `_ndarary_values` represent two different concepts.

> do we get anything from foo._ndarray_values that we don't get from np.array(foo)? (at least in principle assuming __array__ were implemented where relevant)

Yes, cases where you can store type information separate from the values (e.g. writing to HDF5). With `np.array(foo)`, the type information is just dropped.

> are we going to add separate to_arrow down the road or combine that into to_numpy? If the latter, should it have a more generic name?

Possibly, and no I don't think so. The goal of `to_numpy` is to satisfy places that really do need an ndarary.

> does the array property (eventually) replace values/_values? Does/can/should it have any relation to internal_values/external_values/formatting_values?

Yes. IMO we can replace uses of `._values` with `._array` once DatetimeArray is done. I think that we're closeish to removing all the `internal_values` / `external_values` stuff. `formatting_values` will stick around.

> Is it relevant that with DatetimeArray in place we may be able to make .values==._values

My proposal is to not touch `.values` (backwards compat). My unstated goal is to eventually remove uses of it from the pandas docs, and encourage `.array` or `.to_numpy()` depending on the use case.

>  I'd return pd.TimedeltaArray for the timedelta64[ns] case (just barely enough of our behavior differs from numpy's to make that meaningful).

I was hoping to keep `TimedeltaArray` completely invisible to users, since NumPy's `timedelta64[ns]` is sufficient for us. I viewed `TimedeltaArray` as an internal class to make sharing code between DatetimelikeArray easier.

**Comment by TomAugspurger** [comment]
One more edge case: MultiIndex.

Proposal:

- MultiIndex.array -> `ValueError`?
- `MultiIndex.to_numpy() -> `ndarray[object]` (tuples)

I'd raise for `MultiIndex.array`, since there isn't an actual array backing it.

**Comment by jorisvandenbossche** [comment]
I think all those questions about `_ndarray_values` are important to discuss, but not fully relevant for this thread about the *public* API.
(of course if we still want to add something resembling it to the public API, those questions are certianly relevant to understand its usecase, but personally, as argued above, I would not add it to the public API)

As already touched on this in some of the related issues/PRs, the `_ndarray_values` attribute is currently a not well defined part of our EA interface. Not well defined both in what we expect from it as return value, as to where it is then used in the internals. This is certainly worth further figuring out (it also comes up in my WIP ExtensionIndex PR), but maybe in a separate issue.

> I was hoping to keep TimedeltaArray completely invisible to users, since NumPy's timedelta64[ns] is sufficient for us. I viewed TimedeltaArray as an internal class to make sharing code between DatetimelikeArray easier.

Hmm, yeah, I think both arguments make sense. It is indeed not *needed* to return the TimedeltaArray, as an timedelta64[ns] ndarray does not loose any information (although the freq ..), but on the other hand it seems more consistent to return the TimedeltaArray (it feels also more 'future-robust' when we would more and more move towards our own array-likes).

> MultiIndex.array -> ValueError?

+1

> MultiIndex.to_numpy() -> ndarray[object] (tuples)

But then a 1D array? That's certainly consistent with what `np.array(MultiIndex)` does, so the most logical thing to do probably. But it could also be a 2D array similar to np.array(DataFrame). I would also be fine with raising an error as well.

**Comment by TomAugspurger** [comment]
Fair point about TimedeltaArray. I forgot about freq (and the future proofing is nice).

For `MultiIndex.to_numpy()`, I was thinking a 1D array of tuples

```python
In [10]: idx = pd.MultiIndex.from_tuples([('a', 1), ('a', 2)])

In [11]: idx.to_numpy()
Out[11]: array([('a', 1), ('a', 2)], dtype=object)
```

But a 2d array is also reasonable (depending on the expected use case)...

**Comment by jbrockmendel** [comment]
> I think all those questions about _ndarray_values are important to discuss, but not fully relevant for this thread about the public API.

Fair enough.  FWIW I was referring to _downstream_ use cases, not pandas internal usage.

**Comment by TomAugspurger** [comment]
To update the table at 
https://github.com/pandas-dev/pandas/issues/19954#issuecomment-436345938, the proposal is now

dtype              | array                    | to_numpy
------------------ | -------------------------| --------
datetime64[ns]     | DatetimeArray            | ndarray[datetime64[ns]] (view)
datetime64[ns, tz] | DatetimeArray            | ndarray[object] (Timestamp)
timedelta64[ns]    | TimedeltaArray | ndarray[timedelta64[ns]] (view)

which is pleasingly symmetric. OK by everyone?

**Comment by jreback** [comment]
+1 on https://github.com/pandas-dev/pandas/issues/19954#issuecomment-436975784

a ``MultiIndex.to_numpy()`` should be an ndarray of tuples, similar to what ``.values`` does now. If you for example wanted to have an ``expand=True`` kwarg to make that return a 2-d ndarray wouldn't object.

I think the discussion for ``_ndarray_values`` is quite relevant here. There are too many things going on. What exactly are the guarantees that ``._ndarray_values`` has that is not replicable by ``._values``? 

a bit -0 on ``.array`` (property), why not ``.to_array()``, which gives way more flexibility going forward.

**Comment by jorisvandenbossche** [comment]
We indeed need to decide on the *default* behaviour of `to_numpy`? The "maximally information retaining" one (so objects for tz aware datetimes) ? 
In that sense, it is different from `np.array(..)` (`__array__`, which could also be the default behaviour of `to_numpy`), so it might be good to add to that table.

Having the default `to_numpy()` without kwargs be identical to `np.array(..)` gives some consistency which might be nice, but of course puts the distinct behaviour (eg objects, the case for which you might want to use `to_numpy`) behind a keyword.

> which is pleasingly symmetric.  
> [from the table] datetime64[ns, tz] |  

To be clear this is not a numpy dtype (I would make that clearer in the formatting), so in that sense the table is not fully symmetric.

**Comment by TomAugspurger** [comment]
> There are too many things going on.

That's part of why I want to ignore it for now :)

> What exactly are the guarantees that ._ndarray_values has that is not replicable by ._values?

`._values` and `._ndarray_values` are quite different. For e.g. `Series[category]`. `_ndarray_values` is an ndarray of codes. `._values` (now being named `.array`), is the Categorical.

> a bit -0 on .array (property),

People seem to like the property. And if `.array` is supposed to be "the actual array backing this Series" I don't think any keywords make sense. We're just handing you the array, nothing else.

**Comment by TomAugspurger** [comment]
> To be clear this is not a numpy dtype (I would make that clearer in the formatting), so in that sense the table is not fully symmetric.

Changed to DatetimeTZDtype

**Comment by jorisvandenbossche** [comment]
> I think the discussion for _ndarray_values is quite relevant here. There are too many things going on. What exactly are the guarantees that ._ndarray_values has that is not replicable by ._values?

Again, this is discussion is for sure very relevant, and one we need to have. But I think it is orthogonal enough from the discussion about *public* attributes to keep it separate and not further complicate this already complex issue thread.

> a bit -0 on .array (property), why not .to_array(), which gives way more flexibility going forward.

The idea is that `.array` always simply gives the underlying array-like (ndarray or EA) how it is actually stored in the container (series or index). So that *feels* as a property to me, and there is not really a reason to have it as a method?

**Comment by TomAugspurger** [comment]
I add `np.array(...)` to the table, which makes my proposal for `Series[datetime64[ns, tz]]` stand out a bit. `np.array` on that returns the array converted to UTC and tzinfo dropped. We can't change that for backwards compatibility, so should we make `.to_array()` match it?

**Comment by jorisvandenbossche** [comment]
Opened https://github.com/pandas-dev/pandas/issues/23565 for the discussion on `_ndarray_values`

**Comment by jorisvandenbossche** [comment]
>  so should we make .to_array() match it?

@TomAugspurger you mean `to_numpy`?

**Comment by TomAugspurger** [comment]
Yes.

**Comment by jreback** [comment]
> The idea is that .array always simply gives the underlying array-like (ndarray or EA) how it is actually stored in the container (series or index). So that feels as a property to me, and there is not really a reason to have it as a method?

a property just boxes you in as you can never add arguments. ``.to_array()`` is much more friendly.

**Comment by jreback** [comment]
why is ``np.array()`` *any* different than ``.to_numpy()``?

**Comment by jorisvandenbossche** [comment]
But why would you want an argument for it? Can you give a use case? There is only *one* array stored in the series/index container, and it simply returns that one (without making a copy, without any modification) 

>  is np.array() any different than .to_numpy()?

If we do the last change that Tom mentioned above (https://github.com/pandas-dev/pandas/issues/19954#issuecomment-436980762), there might be no difference in the *default* behaviour, but `to_numpy` would have the flexibility of having custom keywords to eg keep timezone information. 
Of course, if that is our only use case, you can also get that with `np.array(.., dtype=object)` ..

Do we have other use cases where multiple ndarrays are possible (except for the datetime tz case) ?

**Comment by jreback** [comment]
> But why would you want an argument for it? Can you give a use case? There is only one array stored in the series/index container, and it simply returns that one (without making a copy, without any modification)

this is exactly the point! ``.values`` boxed us in for some many years, you want to repeat this???

**Comment by jorisvandenbossche** [comment]
> .values boxed us in for some many years, you want to repeat this???

Can you explain what you mean with "boxed us in" ?

**Comment by TomAugspurger** [comment]
I think part of the idea is to *not* have arguments. `.array` is to provide the thing boxed by the series / index, full stop.

> why is np.array() any different than .to_numpy()?

I think we're moving to it matching by default.

> Do we have other use cases where multiple ndarrays are possible (except for the datetime tz case) ?

Not saying this is a good idea, but potentially a 2-D array from an IntervalArray.

**Comment by TomAugspurger** [comment]
Maybe we can pause this conversation until this afternoon :)

**Comment by jreback** [comment]
> If we do the last change that Tom mentioned above (#19954 (comment)), there might be no difference in the default behaviour, but to_numpy would have the flexibility of having custom keywords to eg keep timezone information.
Of course, if that is our only use case, you can also get that with np.array(.., dtype=object) ..

This again offers too many options.

**Comment by jreback** [comment]
> I think we're moving to it matching by default.

great

**Comment by jreback** [comment]
> Can you explain what you mean with "boxed us in" ?

why do you think we have a problem converting Datetime w/tz now? because we can't touch ``.values`` in any way. This is boxed in. We have NO choice but to return a single values no matter what. Imagine if ``.values(tz=True)`` or whatever was available that you would have a lot more flexibility.

A method offers possibilities, a property precludes them.

**Comment by jbrockmendel** [comment]
> .array is to provide the thing boxed by the series / index, full stop.

Suppose we were to implement RangeArray to back RangeIndex (or even range-like DatetimeArray/Index).  Then there is nothing being boxed.  I don't have a real opinion here, just sayin'

**Comment by jorisvandenbossche** [comment]
> Suppose we were to implement RangeArray to back RangeIndex. Then there is nothing being boxed.

Then it is the RangeArray that is being boxed, and thus what would be returned?
